addons/binding/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/handler/NibeHeatPumpHandler.java

```diff
@@ -0,0 +1,532 @@
+        try {
+
+            if (logger.isTraceEnabled()) {
+                logger.trace("Received raw data: {}", msg.toHexString());
+            }
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Received message: {}", msg.toString());
+            }
+
+            updateStatus(ThingStatus.ONLINE);
+
+            if (msg instanceof ModbusReadResponseMessage) {
+                handleReadResponseMessage((ModbusReadResponseMessage) msg);
+            } else if (msg instanceof ModbusWriteResponseMessage) {
+                handleWriteResponseMessage((ModbusWriteResponseMessage) msg);
+            } else if (msg instanceof ModbusDataReadOutMessage) {
+                handleDataReadOutMessage((ModbusDataReadOutMessage) msg);
+            } else {
+                logger.debug("Received unknow message: {}", msg.toString());
+            }
+        } catch (Exception e) {
+            logger.debug("Error occured when parsing recevied message, reason: {}", e.getMessage());
+        }
+    }
+
+    @Override
+    public void errorOccured(String error) {
+        logger.debug("Error '{}' occured, re-establish the connection", error);
+        reconnectionRequest = true;
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, error);
+    }
+
+    private void handleReadResponseMessage(ModbusReadResponseMessage msg) {
+        if (readResult != null) {
+            readResult.set(msg);
+        }
+    }
+
+    private void handleWriteResponseMessage(ModbusWriteResponseMessage msg) {
+        if (writeResult != null) {
+            writeResult.set(msg);
+        }
+    }
+
+    private void handleDataReadOutMessage(ModbusDataReadOutMessage msg) {
+        List<ModbusValue> regValues = msg.getValues();
+
+        if (regValues != null) {
+            for (ModbusValue val : regValues) {
+                handleVariableUpdate(pumpModel, val);
+            }
+        }
+    }
+
+    private void handleVariableUpdate(PumpModel pumpModel, ModbusValue value) {
+        logger.debug("Received variable update: {}", value);
+        int coilAddress = value.getCoilAddress();
+
+        VariableInformation variableInfo = VariableInformation.getVariableInfo(pumpModel, coilAddress);
+
+        if (variableInfo != null) {
+            logger.trace("Using variable information to register {}: {}", coilAddress, variableInfo);
+
+            double val = (double) value.getValue() / (double) variableInfo.factor;
+            logger.debug("{} = {}", coilAddress + ":" + variableInfo.variable, val);
+
+            CacheObject oldValue = stateMap.get(coilAddress);
+            stateMap.put(coilAddress, new CacheObject(System.currentTimeMillis() / 1000, val));
+
+            if (oldValue != null && val == oldValue.value) {
+                logger.trace("Value haven't been changed, ignore update");
+            } else {
+                final String channelPrefix = (variableInfo.type == Type.Setting ? "setting#" : "sensor#");
```

Comment:
Because for local variables their scope should be so short that you always have enough overview so you do not need final to guard the fact that it is not accidentally overwritten. 

For fields which have way broader scope I'm a fan, but for local's it is imho noise. 

Especially when using it inconsistently or on fields which are mutable like `List`s, `array`s, in this case it was one of the few local `final` I saw without any special reason why this is more `final` than all the other local variables.